ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkId.java

```diff
@@ -0,0 +1,378 @@
+        } else {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  private ArrayDeque<ForkIdEntry> collectForksAndHashes(
+      final Set<Long> forks, final Long currentHead) {
+    boolean first = true;
+    ArrayDeque<ForkIdEntry> forkList = new ArrayDeque<>();
+    Iterator<Long> iterator = forks.iterator();
+    while (iterator.hasNext()) {
+      Long forkBlockNumber = iterator.next();
+      if (highestKnownFork < forkBlockNumber) {
+        highestKnownFork = forkBlockNumber;
+      }
+      if (first) {
+        // first fork
+        first = false;
+        forkList.add(
+            new ForkIdEntry(
+                updateCrc(this.genesisHash.getHexString()), forkBlockNumber)); // Genesis
+        updateCrc(forkBlockNumber);
+
+      } else if (!iterator.hasNext()) {
+        // most recent fork
+        forkList.add(new ForkIdEntry(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+        lastKnownEntry = new ForkIdEntry(getCurrentCrcHash(), 0L);
+        forkList.add(lastKnownEntry);
+        if (currentHead > forkBlockNumber) {
+          this.forkNext = 0L;
+        } else {
+          this.forkNext = forkBlockNumber;
+        }
+
+      } else {
+        forkList.add(new ForkIdEntry(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+      }
+    }
+    return forkList;
+  }
+
+  private String updateCrc(final Long block) {
+    byte[] byteRepresentationFork = longToBigEndian(block);
+    crc.update(byteRepresentationFork, 0, byteRepresentationFork.length);
+    return getCurrentCrcHash();
+  }
+
+  private String updateCrc(final String hash) {
+    byte[] byteRepresentation = hexStringToByteArray(hash);
+    crc.update(byteRepresentation, 0, byteRepresentation.length);
+    return getCurrentCrcHash();
+  }
+
+  public String getCurrentCrcHash() {
+    return "0x" + encodeHexString(BytesValues.ofUnsignedInt(crc.getValue()).getByteArray());
+  }
+
+  // TODO use Hash class instead of string for checksum.  convert to or from string only when needed
+  public static class ForkIdEntry {
+    String hash;
+    long next;
+    BytesValue forkIdRLP;
+
+    public ForkIdEntry(final String hash, final long next) {
+      this.hash = hash;
+      this.next = next;
+      createForkIdRLP();
+    }
+
+    public void createForkIdRLP() {
+      BytesValueRLPOutput out = new BytesValueRLPOutput();
+      out.writeList(asList(), ForkId.ForkIdEntry::writeTo);
+      forkIdRLP = out.encoded();
+    }
+
+    public void writeTo(final RLPOutput out) {
+      out.startList();
+      out.writeBytesValue(wrap(hash.getBytes(StandardCharsets.US_ASCII)));
+      out.writeLong(next);
+      out.endList();
+    }
+
+    public List<byte[]> asByteList() {
+      ArrayList<byte[]> forRLP = new ArrayList<byte[]>();
+      forRLP.add(hexStringToByteArray(hash));
+      forRLP.add(longToBigEndian(next));
```

Comment:
It would actually be great to have a `Uint64` type that provides some of this functionality (get a big endian `BytesValue` for instance).

Reply:
ok, cool.  I may be overthinking this, are you thinking such as a 'full' type along the lines of those classes in util.uint (i.e. int256 and Uint256)